2010-09-15  Aurelien Jarno  <aurelien@aurel32.net>
 
	* resolv/res_libc (__res_maybe_init): Ask other threads to
	to re-initialize their resolver states when _PATH_RESCONF
	changes.

--- a/resolv/res_libc.c
+++ b/resolv/res_libc.c
@@ -100,10 +100,13 @@
 
 	if (resp->options & RES_INIT) {
 		ret = stat (_PATH_RESCONF, &statbuf);
-		if ((__res_initstamp != resp->_u._ext.initstamp)
-		    || ((ret == 0) && (last_mtime != statbuf.st_mtime))) {
-			if (ret == 0)
-				last_mtime = statbuf.st_mtime;
+		if ((ret == 0) && (last_mtime != statbuf.st_mtime)) {
+			last_mtime = statbuf.st_mtime;
+			atomicinclock (lock);
+			atomicinc (__res_initstamp);
+			atomicincunlock (lock);
+		}
+		if (__res_initstamp != resp->_u._ext.initstamp) {
 			if (resp->nscount > 0)
 				__res_iclose (resp, true);
 			return __res_vinit (resp, 1);
